modules/csg/csg_shape.cpp

```diff
@@ -503,7 +504,13 @@ void CSGShape3D::_notification(int p_what) {
 			set_collision_mask(collision_mask);
 		}
 
-		_make_dirty();
+		if (!_get_brush() || (parentn && last_parent != Object::cast_to<CSGShape3D>(parentn))) {
+			_make_dirty();
```

Comment:
`_get_brush` returns the current `brush` if not `dirty`. The optimization comes with preventing `_make_dirty` calls.

However, this check was added very early on so it may not still be necessary. I'll review this first thing tomorrow.

Reply:
This check was indeed unnecessary, so it's been removed.